consensus/README.md

```diff
@@ -0,0 +1,205 @@
+
+## Table of Contents <!-- omit in toc -->
+
+- [Interface](#interface)
+- [Consensus Processes](#consensus-processes)
+  - [Leader Election](#leader-election)
+  - [Consensus Phases](#consensus-phases)
+  - [Block Generation](#block-generation)
+  - [Block Validation](#block-validation)
+  - [Consensus Lifecycle](#consensus-lifecycle)
+  - [State Sync](#state-sync)
+- [Implementation](#implementation)
+  - [Code Organization](#code-organization)
+- [Testing](#testing)
+  - [Running Unit Tests](#running-unit-tests)
+
+## Interface
+
+This module adheres to the interface outlined in `pocket/shared/modules/consensus_module.go`, which is derived from the above specification.
+
+## Consensus Processes
+
+This repository features an implementation of the HotStuff consensus algorithm. It facilitates the consensus process through a series of rounds. Staked validator nodes participate in the consensus process, with one node serving as the leader and the rest as replicas.
+
+### Leader Election
+
+A dedicated submodule handles the leader election process. The current configuration employs a deterministic round-robin leader election mechanism.
+
+### Consensus Phases
+
+The HotStuff consensus algorithm consists of three phases: `Prepare`, `Pre-Commit`, and `Commit`. In each phase, the leader creates a proposal and broadcasts it to all replica nodes.
+
+Upon receiving the proposal, each replica node performs a block validation check. If the proposal is valid, the replica node responds to the leader with its signature as a vote.
+
+When the leader collects votes from more than two-thirds of the replicas, it progresses to the next consensus phase. This two-thirds rule is essential to achieve Byzantine Fault Tolerance (BFT), ensuring network resilience against faulty or malicious nodes.
+
+### Block Generation
+
+Block generation in the HotStuff consensus algorithm involves a series of interactive steps between the Leader and the Replica nodes. The steps are as follows:
+
+```mermaid
+sequenceDiagram
+    participant Leader
+    participant Replicas
+    Note over Leader,Replicas: Leader Election
+    Leader->>Replicas: Propose(block)
+    Note over Replicas: Validate proposed block
+    Replicas-->>Leader: Prepare(block)
+    Note over Leader: Receives Prepare messages from a quorum of Replicas
+    Leader->>Replicas: Pre-Commit(block, Prepare messages)
+    Note over Replicas: Validate Pre-Commit message
+    Replicas-->>Leader: Commit(block)
+    Note over Leader: Receives Commit messages from a quorum of Replicas
+    Leader->>Replicas: Notify(block, Commit messages)
+    Note over Replicas: Add block to local blockchain copy
+    Note over Leader,Replicas: New Leader Election
+```
+
+### Block Validation
+
+Every proposal made by the leader goes through a series of validation steps performed by the replicas.
```

Comment:
We should  mentional this only happens in the `Propose` step, and the other steps are just about signature validation. I think you imply it but we should try to make it really explicit.